src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -66,98 +70,66 @@ private class RsFnInferenceContext(
-        psi.forEachChild(this::walk)
+        return block.expr?.let(this::inferExprType) ?: TyUnit
     }
 
-    private fun walk(psi: PsiElement) {
+    fun processStatement(psi: RsStmt) {
         when (psi) {
             is RsLetDecl -> {
-                walkNullable(psi.expr)
-                ctx.extractBindings(psi.pat, psi.typeReference?.type ?: psi.expr?.ty ?: TyUnknown)
+                val exprTy = psi.expr?.let(this::inferExprType) ?: TyUnknown
+                ctx.extractBindings(psi.pat, psi.typeReference?.type ?: exprTy)
             }
-            is RsCondition -> {
-                walkNullable(psi.expr)
-                ctx.extractBindings(psi.pat, psi.expr.ty)
-            }
-            is RsMatchArm -> {
-                for (pat in psi.patList) {
-                    ctx.extractBindings(pat, psi.parentOfType<RsMatchExpr>()?.expr?.ty ?: TyUnknown)
-                }
-                walkChildren(psi)
-            }
-            is RsForExpr -> {
-                walkNullable(psi.expr)
-                ctx.extractBindings(psi.pat, lookup.findIteratorItemType( psi.expr?.ty ?: TyUnknown))
-                psi.block?.let { walkBlock(it) }
-                ctx.writeTy(psi, TyUnit)
-            }
-            is RsLambdaExpr -> {
-                val params = psi.valueParameterList.valueParameterList
-                for (p in params) {
-                    p.typeReference?.type?.let { ctx.extractBindings(p.pat, it) }
-                }
-                val ty = inferTypeForLambdaExpr(psi)
-                if (ty is TyFunction) {
-                    params.zip(ty.paramTypes).forEach { (p, t) -> ctx.extractBindings(p.pat, t) }
-                }
-                walkNullable(psi.expr)
-                ctx.writeTy(psi, if (ty is TyFunction) TyFunction(ty.paramTypes, psi.expr?.ty ?: ty.retType) else TyUnknown)
-            }
-            is RsExpr -> {
-                walkChildren(psi)
-                ctx.writeTy(psi, inferExpressionType(psi))
-            }
-            else -> {
-                walkChildren(psi)
+            is RsExprStmt -> {
+                inferExprType(psi.expr)
             }
         }
     }
 
-    private fun inferExpressionType(expr: RsExpr) = when (expr) {
-        is RsPathExpr -> inferPathExprType(expr)
-        is RsStructLiteral -> inferStructLiteralType(expr)
-        is RsTupleExpr -> TyTuple(expr.exprList.map { it.ty })
-        is RsParenExpr -> expr.expr.ty
-        is RsUnitExpr -> TyUnit
-        is RsCastExpr -> expr.typeReference.type
-        is RsCallExpr -> inferCallExprType(expr)
-        is RsDotExpr -> {
-            val methodCall = expr.methodCall
-            val fieldLookup = expr.fieldLookup
-            when {
-                methodCall != null -> inferMethodCallExprType(methodCall)
-                fieldLookup != null ->inferFieldExprType(fieldLookup)
-                else -> TyUnknown
-            }
+    private fun inferExprType(expr: RsExpr): Ty {
```

Comment:
This could be a local extension property, so that `expr?.inferType :? TyUnknown` works.